### PR TITLE
KTIJ-25503 - Create ClassImportFilter and associated EP

### PR DESCRIPTION
**Pull request by omega697**
This change addresses https://youtrack.jetbrains.com/issue/KTIJ-25503

Currently, some actions like refactoring cause imports to happen
as a side-effect, notably for nested classes. For certain classes,
it is almost never correct to import them. An example here is in
Android Studio where R.string, R.values, R.drawable, etc. are
technically nested classes (despite the lowercase letter).

This change creates a new ClassImportFilter interface and associated
Extension Point, which plugins can use to veto the import of particular
classes.

### CODE REVIEW

File: plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/util/ClassImportFilter.kt [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+import org.jetbrains.kotlin.psi.KtFile
+
+/** Filter that can block classes from being automatically imported as a side-effect of other actions. */
+fun interface ClassImportFilter {
+    /** Returns whether to allow this class to be imported. */
+    fun allowClassImport(descriptor: DeclarationDescriptor, contextFile: KtFile) : Boolean
+
+    companion object {
+        val EP_NAME = ExtensionPointName.create<ClassImportFilter>("org.jetbrains.kotlin.classImportFilter")
+        fun allowClassImport(descriptor: DeclarationDescriptor, contextFile: KtFile) =
```
[the remark this answers is not in the document]
Comment (omega697) [reply plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/util/ClassImportFilter.kt:14]: Is there something equivalent that is FE-agnostic I can use to get the same information?

[the remark this answers is not in the document]
Comment (omega697) [reply plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/util/ClassImportFilter.kt:14]: That, importantly, can be derived from the descriptor in `ImportInsertHelperImpl.kt`?

[the remark this answers is not in the document]
Comment (vladimirdolzhenko) [reply plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/util/ClassImportFilter.kt:14]: Maybe generic FqName would be enough?

[the remark this answers is not in the document]
Comment (omega697) [reply plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/util/ClassImportFilter.kt]: How's this? :)

**Comment by omega697** [comment]
Fixed! Thank you! :)

**Comment by vladimirdolzhenko** [comment]
Thank you for the contribution, merged as https://github.com/JetBrains/intellij-community/commit/a6d67668527b0fc8e4862c789bdbdd7e55425a56

**Pull request closed without merging**